ccbuilder/builder/builder.py

```diff
@@ -138,125 +135,202 @@ def gcc_build_and_install(prefix: Path, cores: int, log_file: TextIO) -> None:
-            llvm_build_and_install(install_path, cores, log_file)
+        match project:
+            case CompilerProject.GCC:
+                gcc_build_and_install(install_path, cores, log_file)
+            case CompilerProject.LLVM:
+                llvm_build_and_install(install_path, cores, log_file)
     except subprocess.CalledProcessError as e:
         raise BuildException(f"Build failed: {e}")
     return install_path
 
 
+def get_install_path(
+    cache_prefix: Path, project: CompilerProject, commit: Commit
+) -> Path:
+    match project:
+        case CompilerProject.LLVM:
+            install_name_prefix = "clang"
+        case CompilerProject.GCC:
+            install_name_prefix = "gcc"
+    return cache_prefix / f"{install_name_prefix}-{commit}"
+
+
+def get_executable_postfix(project: CompilerProject) -> Path:
+    match project:
+        case CompilerProject.LLVM:
+            executable_name = "clang"
+        case CompilerProject.GCC:
+            executable_name = "gcc"
+    return Path("bin") / executable_name
+
+
 def build_and_install_compiler(
-    job: CompilerBuildJob,
-    prefix: Path,
-    cores: int,
-    additional_patches: list[Path] = [],
+    project: CompilerProject,
+    rev: Revision,
+    cache_prefix: Path,
+    llvm_repo: Repo,
+    gcc_repo: Repo,
+    patchdb: PatchDB,
+    get_executable: bool = False,
+    jobs: Optional[int] = None,
+    additional_patches: Optional[list[Path]] = None,
     logdir: Optional[Path] = None,
 ) -> Path:
-    install_path = get_install_path_from_job(job, prefix)
+
+    repo = select_repo(project, llvm_repo, gcc_repo)
+    commit = repo.rev_to_commit(rev)
+    install_path = get_install_path(cache_prefix, project, commit)
     success_indicator = install_path / "DONE"
 
     if not logdir:
-        logdir = prefix / "logs"
+        logdir = cache_prefix / "logs"
     logdir.mkdir(exist_ok=True)
-    with BuildContext(install_path, success_indicator, job, logdir) as (
+    with BuildContext(cache_prefix, success_indicator, project, commit, logdir) as (
         tmpdir,
         build_log,
     ):
         run_cmd_to_logfile(
-            f"git -C {str(job.compiler_config.repo.path)} worktree"
-            f" add {tmpdir} {job.commit_to_build} -f",
+            f"git -C {str(repo.path)} worktree" f" add {tmpdir} {commit} -f",
             log_file=build_log,
         )
-        patch_if_necessary(job, Path(tmpdir), additional_patches)
-        res = _run_build_and_install(job, prefix, cores, build_log)
+        patch_if_necessary(
+            commit,
+            patchdb,
+            Path(tmpdir),
+            additional_patches if additional_patches else [],
+        )
+        res = _run_build_and_install(
+            project,
+            install_path,
+            jobs if jobs else multiprocessing.cpu_count(),
+            build_log,
+        )
         success_indicator.touch()
+        if get_executable:
```

Comment:
Isn't this a bit strange? To change the return based on the bool argument? Can't this be done at the callsite?

Reply:
Surely it can be done at the callsite but you'll have to do a   
```python
res = builder.build(...)
project = name_to_project(name)
match project:
     case CompilerProject.GCC:
              res = res / "bin" / "gcc"
     case CompilerProject.LLVM:
             res = res / "bin" /  "clang"
```
several times across the code for different situations which becomes annoying quickly.